shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -33,22 +34,37 @@
 
   @NonNull private final FlutterJNI flutterJNI;
 
-  @NonNull private final ConcurrentHashMap<String, HandlerInfo> messageHandlers;
+  /**
+   * Maps a channel name to an object that contains the task queue and the handler associated with
+   * the channel.
+   *
+   * <p>Reads and writes to this map must lock {@code handlersLock}.
+   */
+  @NonNull private final Map<String, HandlerInfo> messageHandlers = new HashMap<>();
+
+  /**
+   * Maps a channel name to an object that holds information about the incoming Dart message.
+   *
+   * <p>Reads and writes to this map must lock {@code handlersLock}.
+   */
+  @NonNull private final Map<String, List<BufferedMessageInfo>> bufferedMessages = new HashMap<>();
 
-  @NonNull private final Map<Integer, BinaryMessenger.BinaryReply> pendingReplies;
+  @NonNull private final Object handlersLock = new Object();
+  private boolean enableBufferingIncomingMessages = false;
```

Comment:
Make this an `AtomicBoolean`
(it will be set on the platform thread and read on the UI thread)

Reply:
right on